core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -474,72 +454,72 @@ public void run() {
-            );
-      submitBatchIfNecessary();
-      return asyncOrThrottledStage();
-   }
-
-   @Override
-   public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> entry) {
-      assertNotStopped();
-      submissionFlowable.onNext(new PutModification(segment, entry));
-      submitBatchIfNecessary();
-      return asyncOrThrottledStage();
-   }
-
-   @Override
-   public CompletionStage<Boolean> delete(int segment, Object key) {
-      assertNotStopped();
-      submissionFlowable.onNext(new RemoveModification(segment, key));
-      submitBatchIfNecessary();
-      return asyncOrThrottledStage()
-            // We always assume it was removed with async
-            .thenCompose(ignore -> CompletableFutures.completedTrue());
-   }
-
-   private void submitBatchIfNecessary() {
+      Maybe<CompletionStage<Void>> maybeRemove = Flowable.fromPublisher(removePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp)
+                  .map(key -> submitModification(new RemoveModification(sp.getSegment(), key))), publisherCount)
+            .lastElement();
+      Maybe<CompletionStage<Void>> maybeModify = Flowable.fromPublisher(writePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp)
+                  .map(me -> submitModification(new PutModification(sp.getSegment(), me))), publisherCount)
+            .lastElement();
+      return maybeRemove.mergeWith(maybeModify)
+            .lastStage(CompletableFutures.completedNull())
+            .thenCompose(java.util.function.Function.identity());
+   }
+
+   CompletionStage<Void> submitModification(Modification modification) {
       boolean startNewBatch;
+      CompletionStage<Void> stageToWaitFor;
       synchronized (this) {
+         modification.apply(this);
+
          if (startNewBatch = batchFuture == null) {
             batchFuture = new CompletableFuture<>();
          }
+         int replicatingSize = replicatingModifications == null ? 0 : replicatingModifications.size();
+
+         stageToWaitFor = pendingModifications.size() + replicatingSize > modificationQueueSize
+               ? batchFuture : null;
+         if (stageToWaitFor != null) {
+            System.currentTimeMillis();
+         }
+      }
+
+      if (trace) {
+         log.tracef("A new modification %s has been enqueued with async store", modification);
       }
 
       if (startNewBatch) {
          if (trace) {
             log.tracef("Requesting a new async batch operation to be ran!");
          }
-         // Any old object will work
-         requestFlowable.onNext(requestFlowable);
+         submitTask();
+      }
+      if (stageToWaitFor != null && trace) {
+         log.tracef("Operation will not return immediately, must wait until current batch completes");
       }
+      return stageToWaitFor == null ? CompletableFutures.completedNull() :
+            stageToWaitFor.thenApplyAsync(CompletableFutures.toNullFunction(), nonBlockingExecutor);
```

Comment:
The problem isn't what thread it is invoked on but rather that a bunch of these could be enqueued waiting for the completion of this stage. The `stageToWaitFor` is the `batchFuture` so you could have hundreds or thousands depending upon this one stage, so we may not want to block completion of all of these on a single thread. But maybe that is okay?

Reply:
I see. 

I was thinking running them on a single thread might be better for throughput, because you have less context switching. But submitting them to the executor is potentially better for latency, so let's keep it this way unless it comes up in the profiler at some point in the future.